Bottom: caef5e458f20125f24964be83a7eaaa6b04b94cf
Top:    211e9d2aa8ee1df6f822eb0cb0b51225177304a9
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-27 23:39:57 +0000

More caching for STEP output


---

diff --git a/src/hid/step/step_writer.c b/src/hid/step/step_writer.c
index 784145a..7c9605d 100644
--- a/src/hid/step/step_writer.c
+++ b/src/hid/step/step_writer.c
@@ -72,6 +72,10 @@ step_file
   file->cartesian_point_hash = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
   file->direction_hash =       g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
   file->vector_hash =          g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  file->axis2_hash =           g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  file->colour_hash =          g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  file->cylindrical_hash =     g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  file->circle_hash =          g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
 
   return file;
 }
@@ -82,6 +86,10 @@ destroy_step_output_file (step_file *file)
   g_hash_table_destroy (file->cartesian_point_hash);
   g_hash_table_destroy (file->direction_hash);
   g_hash_table_destroy (file->vector_hash);
+  g_hash_table_destroy (file->axis2_hash);
+  g_hash_table_destroy (file->colour_hash);
+  g_hash_table_destroy (file->cylindrical_hash);
+  g_hash_table_destroy (file->circle_hash);
   g_free (file);
 }
 
@@ -104,6 +112,23 @@ make_3d_metric_step_geometric_representation_context (step_file *file)
   return file->next_id++;
 }
 
+static step_id
+lookup_or_create (step_file *file, GHashTable *hash, const char *entity, char *content)
+{
+  step_id id;
+
+  if ((id = GPOINTER_TO_INT (g_hash_table_lookup (hash, content))) != 0)
+    {
+      g_free (content);
+      return id;
+    }
+
+  g_hash_table_insert (hash, content, GINT_TO_POINTER (file->next_id));
+  fprintf (file->f, "#%i=%s(%s);\n", file->next_id, entity, content);
+
+  return file->next_id++;
+}
+
 step_id
 step_application_context (step_file *file, char *application)
 {
@@ -196,49 +221,22 @@ step_product_definition_shape (step_file *file, char *name, char *description, s
 step_id
 step_cartesian_point (step_file *file, char *name, double x, double y, double z)
 {
-  char *line;
-  step_id id;
-
-  line = g_strdup_printf ("'%s',(%f,%f,%f)", name, x, y, z);
-
-  if ((id = GPOINTER_TO_INT (g_hash_table_lookup (file->cartesian_point_hash, line))) != 0)
-    {
-      g_free (line);
-      return id;
-    }
-
-  g_hash_table_insert (file->cartesian_point_hash, line, GINT_TO_POINTER (file->next_id));
-  fprintf (file->f, "#%i=CARTESIAN_POINT(%s);\n", file->next_id, line);
-
-  return file->next_id++;
+  char *content = g_strdup_printf ("'%s',(%f,%f,%f)", name, x, y, z);
+  return lookup_or_create (file, file->cartesian_point_hash, "CARTESIAN_POINT", content);
 }
 
 step_id
 step_direction (step_file *file, char *name, double x, double y, double z)
 {
-  char *line;
-  step_id id;
-
-  line = g_strdup_printf ("'%s',(%f,%f,%f)", name, x, y, z);
-
-  if ((id = GPOINTER_TO_INT (g_hash_table_lookup (file->direction_hash, line))) != 0)
-    {
-      g_free (line);
-      return id;
-    }
-
-  g_hash_table_insert (file->direction_hash, line, GINT_TO_POINTER (file->next_id));
-  fprintf (file->f, "#%i=DIRECTION(%s);\n", file->next_id, line);
-
-  return file->next_id++;
+  char *content = g_strdup_printf ("'%s',(%f,%f,%f)", name, x, y, z);
+  return lookup_or_create (file, file->direction_hash, "DIRECTION", content);
 }
 
 step_id
 step_axis2_placement_3d (step_file *file, char *name, step_id location, step_id axis, step_id ref_direction)
 {
-  fprintf (file->f, "#%i=AXIS2_PLACEMENT_3D('%s',#%i,#%i,#%i);\n",
-                    file->next_id, name, location, axis, ref_direction);
-  return file->next_id++;
+  char *content =  g_strdup_printf ("'%s',#%i,#%i,#%i", name, location, axis, ref_direction);
+  return lookup_or_create (file, file->axis2_hash, "AXIS2_PLACEMENT_3D", content);
 }
 
 step_id
@@ -252,37 +250,22 @@ step_plane (step_file *file, char *name, step_id position)
 step_id
 step_cylindrical_surface (step_file *file, char *name, step_id position, double radius)
 {
-  fprintf (file->f, "#%i=CYLINDRICAL_SURFACE('%s',#%i,%f);\n",
-                    file->next_id, name, position, radius);
-  return file->next_id++;
+  char *content = g_strdup_printf ("'%s',#%i,%f", name, position, radius);
+  return lookup_or_create (file, file->cylindrical_hash, "CYLINDRICAL_SURFACE", content);
 }
 
 step_id
 step_circle (step_file *file, char *name, step_id position, double radius)
 {
-  fprintf (file->f, "#%i=CIRCLE('%s',#%i,%f);\n",
-                    file->next_id, name, position, radius);
-  return file->next_id++;
+  char *content = g_strdup_printf ("'%s',#%i,%f", name, position, radius);
+  return lookup_or_create (file, file->circle_hash, "CIRCLE", content);
 }
 
 step_id
 step_vector (step_file *file, char *name, step_id orientation, double magnitude)
 {
-  char *line;
-  step_id id;
-
-  line = g_strdup_printf ("'%s',#%i,%f", name, orientation, magnitude);
-
-  if ((id = GPOINTER_TO_INT (g_hash_table_lookup (file->vector_hash, line))) != 0)
-    {
-      g_free (line);
-      return id;
-    }
-
-  g_hash_table_insert (file->vector_hash, line, GINT_TO_POINTER (file->next_id));
-  fprintf (file->f, "#%i=VECTOR(%s);\n", file->next_id, line);
-
-  return file->next_id++;
+  char *content = g_strdup_printf ("'%s',#%i,%f", name, orientation, magnitude);
+  return lookup_or_create (file, file->vector_hash, "VECTOR", content);
 }
 
 step_id
@@ -398,9 +381,8 @@ step_shape_definition_representation (step_file *file, step_id definition, step_
 step_id
 step_colour_rgb (step_file *file, char *name, double red, double green, double blue)
 {
-  fprintf (file->f, "#%i=COLOUR_RGB('%s',%f,%f,%f);\n",
-                    file->next_id, name, red, green, blue);
-  return file->next_id++;
+  char *content = g_strdup_printf ("'%s',%f,%f,%f", name, red, green, blue);
+  return lookup_or_create (file, file->colour_hash, "COLOUR_RGB", content);
 }
 
 step_id
diff --git a/src/hid/step/step_writer.h b/src/hid/step/step_writer.h
index 1702b77..3fe3ca4 100644
--- a/src/hid/step/step_writer.h
+++ b/src/hid/step/step_writer.h
@@ -8,7 +8,10 @@ typedef struct {
   GHashTable *cartesian_point_hash;
   GHashTable *direction_hash;
   GHashTable *vector_hash;
-
+  GHashTable *axis2_hash;
+  GHashTable *colour_hash;
+  GHashTable *cylindrical_hash;
+  GHashTable *circle_hash;
 } step_file;
 
 step_id_list step_id_list_append (step_id_list list, step_id id);
